File: client/ios/Data/Catalog/CatalogHttpGateway.swift
import Foundation

public final class CatalogHttpGateway: CatalogBrowseGateway, CatalogNodeDetailGateway, @unchecked Sendable {
    private let baseUrl: String
    private let catalogCode: String
    private let session: URLSession

    public init(baseUrl: String, catalogCode: String, session: URLSession = .shared) {
        self.baseUrl = baseUrl
        self.catalogCode = catalogCode
        self.session = session
    }

    public func listNodes(query: ListCatalogNodesQuery) async throws -> [CatalogNodeSummary] {
        var components = URLComponents(string: baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/catalog")
        var items = [URLQueryItem(name: "catalogCode", value: query.catalogCode?.isEmpty == false ? query.catalogCode : catalogCode)]
        if let parentNodeId = query.parentNodeId, !parentNodeId.isEmpty {
            items.append(URLQueryItem(name: "parentNodeId", value: parentNodeId))
        }
        if let searchText = query.searchText, !searchText.isEmpty {
            items.append(URLQueryItem(name: "q", value: searchText))
        }
        components?.queryItems = items
        guard let url = components?.url else { throw URLError(.badURL) }

        let payload = try await get(url)
        return (payload["nodes"] as? [[String: Any]] ?? []).compactMap(Self.node)
    }

    public func loadNodeDetail(nodeId: String) async throws -> CatalogNodeDetailPayload {
        let encoded = nodeId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? nodeId
        guard let url = URL(string: baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/catalog/node/\(encoded)") else {
            throw URLError(.badURL)
        }
        let payload = try await get(url)
        guard let nodePayload = payload["node"] as? [String: Any], let node = Self.node(nodePayload) else {
            throw URLError(.cannotParseResponse)
        }

        return CatalogNodeDetailPayload(
            node: node,
            description: payload["description"] as? String,
            breadcrumbLabels: payload["breadcrumbLabels"] as? [String] ?? [],
            featuredProductIds: payload["featuredProductIds"] as? [String] ?? []
        )
    }

    private func get(_ url: URL) async throws -> [String: Any] {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        let payload = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        guard (200..<300).contains(http.statusCode) else {
            throw NSError(
                domain: "Catalog",
                code: http.statusCode,
                userInfo: [NSLocalizedDescriptionKey: payload["message"] as? String ?? "Catalog request failed."]
            )
        }
        return payload
    }

    private static func node(_ payload: [String: Any]) -> CatalogNodeSummary? {
        guard let nodeId = string(payload["nodeId"]), !nodeId.isEmpty else { return nil }
        return CatalogNodeSummary(
            nodeId: nodeId,
            parentNodeId: string(payload["parentNodeId"]),
            title: string(payload["title"]) ?? "Catalog item",
            slug: string(payload["slug"]),
            imageUrl: string(payload["imageUrl"]),
            childCount: integer(payload["childCount"]),
            productCount: integer(payload["productCount"])
        )
    }

    private static func string(_ value: Any?) -> String? {
        if let value = value as? String, !value.isEmpty { return value }
        if let value = value as? NSNumber { return value.stringValue }
        return nil
    }

    private static func integer(_ value: Any?) -> Int {
        if let value = value as? NSNumber { return value.intValue }
        if let value = value as? String, let integer = Int(value) { return integer }
        return 0
    }
}

File: client/android/Data/Catalog/CatalogHttpGateway.kt
package app.mobiling.client.data.catalog

import app.mobiling.client.contract.catalog.browse.CatalogListNodeQuery
import app.mobiling.client.contract.catalog.browse.CatalogNodeSummary
import app.mobiling.client.contract.catalog.detail.CatalogNodeDetailPayload
import app.mobiling.client.data.catalog.browse.CatalogBrowseGateway
import app.mobiling.client.data.catalog.detail.CatalogNodeDetailGateway
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

class CatalogHttpGateway(
    private val baseUrl: String,
    private val catalogCode: String,
    private val client: OkHttpClient = OkHttpClient(),
) : CatalogBrowseGateway, CatalogNodeDetailGateway {
    override suspend fun listNodes(query: CatalogListNodeQuery): List<CatalogNodeSummary> = withContext(Dispatchers.IO) {
        val url = (baseUrl.trimEnd('/') + "/catalog").toHttpUrl().newBuilder()
            .addQueryParameter("catalogCode", query.catalogCode?.takeIf(String::isNotBlank) ?: catalogCode)
            .apply {
                query.parentNodeId?.takeIf(String::isNotBlank)?.let { addQueryParameter("parentNodeId", it) }
                query.searchText?.takeIf(String::isNotBlank)?.let { addQueryParameter("q", it) }
            }
            .build()
        val root = get(url.toString())
        parseNodes(root.optJSONArray("nodes") ?: JSONArray())
    }

    override suspend fun loadNodeDetail(nodeId: String): CatalogNodeDetailPayload = withContext(Dispatchers.IO) {
        val root = get(baseUrl.trimEnd('/') + "/catalog/node/" + java.net.URLEncoder.encode(nodeId, Charsets.UTF_8.name()))
        val node = parseNode(root.optJSONObject("node") ?: JSONObject())
        CatalogNodeDetailPayload(
            node = node,
            description = root.optString("description").takeIf(String::isNotBlank),
            breadcrumbLabels = root.optJSONArray("breadcrumbLabels").strings(),
            featuredProductIds = root.optJSONArray("featuredProductIds").strings(),
        )
    }

    private fun get(url: String): JSONObject {
        client.newCall(Request.Builder().url(url).header("Accept", "application/json").get().build()).execute().use { response ->
            val body = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                val message = runCatching { JSONObject(body).optString("message") }.getOrNull()
                throw IllegalStateException(message?.takeIf(String::isNotBlank) ?: "Catalog request failed with HTTP ${response.code}.")
            }
            return JSONObject(body.ifBlank { "{}" })
        }
    }

    private fun parseNodes(array: JSONArray): List<CatalogNodeSummary> = buildList {
        for (index in 0 until array.length()) array.optJSONObject(index)?.let { add(parseNode(it)) }
    }

    private fun parseNode(item: JSONObject) = CatalogNodeSummary(
        nodeId = item.optString("nodeId"), parentNodeId = item.optString("parentNodeId").takeIf(String::isNotBlank),
        title = item.optString("title", "Catalog item"), slug = item.optString("slug").takeIf(String::isNotBlank),
        imageUrl = item.optString("imageUrl").takeIf(String::isNotBlank),
        childCount = item.optInt("childCount", 0), productCount = item.optInt("productCount", 0),
    )
}

private fun JSONArray?.strings(): List<String> = if (this == null) {
    emptyList()
} else {
    buildList {
        for (index in 0 until length()) {
            optString(index).takeIf(String::isNotBlank)?.let(::add)
        }
    }
}
